lib/rules.js

```diff
@@ -353,6 +353,267 @@ export const GROUP_TERMS_BY_ROOT = defineRule(
+        }
+    }
+
+    return factors
+}
+
+// TODO: handle case when exponent is a polynomial (x^b^4)
+// e.g. 12x^2 + 17x + 6 -> (3 x^1 + 2) (4 x^1 + 3)
+export const FACTOR_SUM_PRODUCT_RULE = defineRule(
+    (node) => {
+        if (query.isAdd(node) && node.args.length === 3) {
+            const [firstTerm, secondTerm, thirdTerm] = node.args
+
+            // First two terms should be polynomials
+            // Third term is either number or polynomial
+            const isTrinomial =
+                  isPolynomialTerm(firstTerm)
+                  && isPolynomialTerm(secondTerm)
+                  && (query.isNumber(thirdTerm) || isPolynomialTerm(thirdTerm))
+
+            if (isTrinomial) {
+                let isFactorable = false
+                const {constants, coefficientMap} = getCoefficientsAndConstants(node)
+                const variables = Object.keys(coefficientMap).map(key => parse(key))
+                const coefficients = Object.keys(coefficientMap).map(key => {return coefficientMap[key]})
+
+                const firstPoly = query.isMul(variables[0]) ? variables[0].args : [variables[0]]
+                const thirdPoly =
+                      variables[2]
+                      ? query.isMul(variables[2])
+                      ? variables[2].args
+                      : [variables[2]]
+                      : null
+
+                const secondCoef = coefficients[1][0]
+                // Second polynomial to be matched
+                const firstMiddle = firstPoly.map(term => {
+                    const identifier = term.args[0]
+                    const exponent = build.number(query.getValue(term.args[1]) / 2)
+                    return build.applyNode('pow', [identifier, exponent])
+                })
+
+                let secondMiddle
+                if (thirdPoly) {
+                    secondMiddle = thirdPoly.map(term => {
+                        const identifier = term.args[0]
+                        const exponent = build.number(query.getValue(term.args[1]) / 2)
+                        return build.applyNode('pow', [identifier, exponent])
+                    })
+                }
+
+                const matchSecondTerm =  secondMiddle
+                      ? build.applyNode ('mul',
+                                         [secondCoef, ...firstMiddle, ...secondMiddle]
+                                         , {implicit: true})
+                      : build.applyNode ('mul',
+                                         [secondCoef, ...firstMiddle]
+                                         , {implicit: true})
+
+                // General conditions:
+                // All degrees in 1st and 3rd term must be even
+                // secondTerm = matchSecondTerm
+                isFactorable =
+                    firstPoly.forEach(term => {return term.args[1] % 2 == 0})
+                    && thirdPoly ? thirdPoly.forEach(term => {return term.args[1] % 2 == 0}) : true
```

Comment:
The `return` isn't necessary here.  Also, I think you meant `every` not `forEach`.